### PR TITLE
HDI add autoscale feature support

### PR DESCRIPTION
**Pull request by gaoxiaoninghit**
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [x] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [x] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

**Review by idear1203 — APPROVED** [review]
LGTM.

File: specification/hdinsight/resource-manager/Microsoft.HDInsight/preview/2015-03-01-preview/cluster.json [hunk 1]
```diff
@@ -523,931 +523,1020 @@
...
+          "description": "The minimum instance count of the cluster."
+        },
+        "targetInstanceCount": {
+          "type": "integer",
+          "format": "int32",
+          "description": "The instance count of the cluster."
+        },
+        "autoscale": {
+          "$ref": "#/definitions/Autoscale",
+          "x-ms-client-name": "autoscaleConfiguration",
```
Comment (NelsonDaniel) [inline specification/hdinsight/resource-manager/Microsoft.HDInsight/preview/2015-03-01-preview/cluster.json]: This extension goes inside the schema definition of Autoscale.

File: specification/hdinsight/resource-manager/Microsoft.HDInsight/stable/2018-06-01-preview/cluster.json [hunk 2]
```diff
@@ -523,930 +523,1019 @@
...
-            },
-            "description": "The list runtime script action detail response."
+        "targetInstanceCount": {
+          "type": "integer",
+          "format": "int32",
+          "description": "The instance count of the cluster."
+        },
+        "autoscale": {
+          "$ref": "#/definitions/Autoscale",
+          "x-ms-client-name": "autoscaleConfiguration",
```
Comment (NelsonDaniel) [inline specification/hdinsight/resource-manager/Microsoft.HDInsight/stable/2018-06-01-preview/cluster.json]: This extension goes inside the schema definition of Autoscale.

**Review by NelsonDaniel — CHANGES_REQUESTED** [review]
I left some review comments. Also, solve merge conflicts

File: specification/hdinsight/resource-manager/Microsoft.HDInsight/preview/2015-03-01-preview/cluster.json [hunk 3]
```diff
@@ -523,931 +523,1020 @@
...
-                }
-            },
-            "description": "The storage Account."
+        "msiResourceId": {
+          "type": "string",
+          "description": "User assigned identity that has permissions to read and create cluster-related artifacts in the user's AADDS."
+        }
+      }
+    },
+    "AutoscaleTimeAndCapacity": {
```
[the remark this answers is not in the document]
Comment (gaoxiaoning) [reply specification/hdinsight/resource-manager/Microsoft.HDInsight/preview/2015-03-01-preview/cluster.json]: Refer to other class definitions in `Role`, Such as `virtualNetworkProfile`, `osProfile`,`HardwareProfile`, defining a object do not need or require declaring the type of "object"!

File: specification/hdinsight/resource-manager/Microsoft.HDInsight/preview/2015-03-01-preview/cluster.json [hunk 4]
```diff
@@ -523,931 +523,1020 @@
...
-                }
+        "maxInstanceCount": {
+          "type": "integer",
+          "format": "int32",
+          "description": "The maximum instance count of the cluster"
+        }
+      },
+      "description": "Time and capacity request parameters"
+    },
+    "AutoscaleSchedule": {
```
[the remark this answers is not in the document]
Comment (gaoxiaoning) [reply specification/hdinsight/resource-manager/Microsoft.HDInsight/preview/2015-03-01-preview/cluster.json]: Refer to other class definitions in `Role`, Such as `virtualNetworkProfile`, `osProfile`,`HardwareProfile`, defining a object do not need or require declaring the type of "object"!

File: specification/hdinsight/resource-manager/Microsoft.HDInsight/stable/2018-06-01-preview/cluster.json (hunk 2 above)
Comment (gaoxiaoning) [reply specification/hdinsight/resource-manager/Microsoft.HDInsight/stable/2018-06-01-preview/cluster.json]: Refer to other class definitions in `Role`, Such as `virtualNetworkProfile`, `osProfile`,`HardwareProfile`, We can make this extension goes inside the schema definition of Autoscale. same as definitions  of `osProfile`,`HardwareProfile`

File: specification/hdinsight/resource-manager/Microsoft.HDInsight/preview/2015-03-01-preview/cluster.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (NelsonDaniel) [reply specification/hdinsight/resource-manager/Microsoft.HDInsight/preview/2015-03-01-preview/cluster.json]: It is not a required field, but it ensures there's correct disambiguation for tooling.

File: specification/hdinsight/resource-manager/Microsoft.HDInsight/stable/2018-06-01-preview/cluster.json (hunk 2 above)
Comment (NelsonDaniel) [reply specification/hdinsight/resource-manager/Microsoft.HDInsight/stable/2018-06-01-preview/cluster.json]: For a property, referencing a definition, all the fields should be inside the definition being referenced. In this case, the extension should be inside Autoscale definition. We allow description to be at the same level, but just for documentation purposes.

**Comment by NelsonDaniel** [comment]
Merge conflicts should be resolved.

File: specification/hdinsight/resource-manager/Microsoft.HDInsight/preview/2015-03-01-preview/cluster.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (gaoxiaoning) [reply specification/hdinsight/resource-manager/Microsoft.HDInsight/preview/2015-03-01-preview/cluster.json]: @NelsonDaniel In order to be consistent with other class definitions in Role，we decided not to add   "type" field.

File: specification/hdinsight/resource-manager/Microsoft.HDInsight/stable/2018-06-01-preview/cluster.json (hunk 2 above)
Comment (gaoxiaoning) [reply specification/hdinsight/resource-manager/Microsoft.HDInsight/stable/2018-06-01-preview/cluster.json]: @NelsonDaniel We did an experiment：when we make the extension `"x-ms-client-name": "autoscaleConfiguration"` inside Autoscale definition, the generated code does not match what we expected，so we decide to keep the origin writing style.

Comment (NelsonDaniel) [reply specification/hdinsight/resource-manager/Microsoft.HDInsight/stable/2018-06-01-preview/cluster.json]: Got it. Solve merge conflicts. Apart from that LGTM

**Comment by NelsonDaniel** [comment]
Looks good to me. Just waiting for ARM feedback. Need to reassign this PR since I will be OOF, and this needs to be merged before I come back

**Review by NelsonDaniel — APPROVED** [review]

File: specification/hdinsight/resource-manager/Microsoft.HDInsight/preview/2015-03-01-preview/cluster.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (amarzavery) [reply specification/hdinsight/resource-manager/Microsoft.HDInsight/preview/2015-03-01-preview/cluster.json]: @gaoxiaoning - If you haven't added `"type": "object"` at other places, then please add it in other models and this model definition as well.

[the remark this answers is not in the document]
Comment (amarzavery) [reply specification/hdinsight/resource-manager/Microsoft.HDInsight/preview/2015-03-01-preview/cluster.json]: One cannot vouch for consistency by ensuring the same bug exists in multiple places and introduce the bug while adding new stuff. The correct approach is to fix the bug everywhere.

Please read [this](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/creating-swagger.md#understanding-the-importance-of-type-keyword-while-defining-model-types) to understand the significance of adding `"type": "object"` to a model definition.

**Comment by amarzavery** [comment]
@gaoxiaoning - New model types have been added to the swagger specification. I don't see any updates to the corresponding x-ms-examples in the operation where these types are being used. Please update the existing examples or add new examples that show the usage of newly added model types.

**Comment by amarzavery** [comment]
@gaoxiaoning - New model types have been added to the swagger specification. I don't see any updates to the corresponding x-ms-examples in the operation where these types are being used. Please update the existing examples or add new examples that show the usage of newly added model types.

**Comment by amarzavery** [comment]
@gaoxiaoning - New model types have been added to the swagger specification. I don't see any updates to the corresponding x-ms-examples in the operation where these types are being used. Please update the existing examples or add new examples that show the usage of newly added model types.

**Review by majastrz — APPROVED** [review]
Looks good to me.

**Comment by idear1203** [comment]
@amarzavery , could you please help merge the PR or let us know if you have other concerns:)

**Review by amarzavery — APPROVED** [review]

**Pull request merged**